[RPC] Infer backend type if only options are given

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45065 [RPC] Infer backend type if only options are given**

To preserve backwards compatibility with applications that were passing in some ProcessGroupRpcBackendOptions but were not explicitly setting backend=BackendType.PROCESS_GROUP, we're here now inferring the backend type from the options if only the latter ones are passed. If neither are passed, we'll default to TensorPipe, as before this change.

Differential Revision: [D23814289](https://our.internmc.facebook.com/intern/diff/D23814289/)

lw committed:
[RPC] Infer backend type if only options are given

To preserve backwards compatibility with applications that were passing in some ProcessGroupRpcBackendOptions but were not explicitly setting backend=BackendType.PROCESS_GROUP, we're here now inferring the backend type from the options if only the latter ones are passed. If neither are passed, we'll default to TensorPipe, as before this change.

Differential Revision: [D23814289](https://our.internmc.facebook.com/intern/diff/D23814289/)

[ghstack-poisoned]

lw commented on torch/distributed/rpc/__init__.py:
```diff
@@ -71,7 +76,38 @@ def init_rpc(
+                )
+            if backend != BackendType.TENSORPIPE:
+                logger.warning(
+                    f"RPC was initialized with no explicit backend but with options "
+                    f"corresponding to {backend}, hence that backend will be used "
+                    f"instead of the default {BackendType.TENSORPIPE}. To silence this "
+                    f"warning pass `backend={backend}` explicitly."
```
Should we also add a deprecation warning about ProcessGroup since our long-term plan is to remove it?

pritamdamania87 replied:
Yes, it would be good to add a deprecation warning too.

pritamdamania87 commented on torch/distributed/rpc/__init__.py:
```diff
@@ -71,7 +76,38 @@ def init_rpc(
+        if backend is None:
+            backend = BackendType.TENSORPIPE
+
+        if rpc_backend_options is None:
```
What happens if backend is specified as TensorPipe but backend options are ProcessGroupOptions? Do we raise an appropriate error? If so, can we have a unit test that verifies this?

pritamdamania87 commented on torch/distributed/rpc/__init__.py:
```diff
@@ -71,7 +76,38 @@ def init_rpc(
+                        )
+                    ),
+                ):
+                    backend = candidate_backend
+                    break
```
Are these codecov warnings false positives? It does seem like your test should cover this line?

lw replied:
That's unclear. I'm assuming that the coverage is collected from this job: https://app.circleci.com/pipelines/github/pytorch/pytorch/217281/workflows/3cfc25f2-225c-4be1-9ad5-a7942dc829e7/jobs/7659664 (since it's the only one with `coverage` in the name). And I checked the logs and the tests are run.

One option is that the odd on-the-fly class construction trick we're doing to assemble the RPC test suites makes it hard for the coverage tool to map the test back to its original lines.

Or I misunderstood how this works.

pritamdamania87 commented on torch/distributed/rpc/__init__.py:
```diff
@@ -71,7 +76,38 @@ def init_rpc(
+                    backend = candidate_backend
+                    break
+            else:
+                raise TypeError(
```
Can we add a test covering this?

lw committed:
Update on "[RPC] Infer backend type if only options are given"


To preserve backwards compatibility with applications that were passing in some ProcessGroupRpcBackendOptions but were not explicitly setting backend=BackendType.PROCESS_GROUP, we're here now inferring the backend type from the options if only the latter ones are passed. If neither are passed, we'll default to TensorPipe, as before this change.

Differential Revision: [D23814289](https://our.internmc.facebook.com/intern/diff/D23814289/)

[ghstack-poisoned]

lw committed:
Update on "[RPC] Infer backend type if only options are given"


To preserve backwards compatibility with applications that were passing in some ProcessGroupRpcBackendOptions but were not explicitly setting backend=BackendType.PROCESS_GROUP, we're here now inferring the backend type from the options if only the latter ones are passed. If neither are passed, we'll default to TensorPipe, as before this change.

Differential Revision: [D23814289](https://our.internmc.facebook.com/intern/diff/D23814289/)

[ghstack-poisoned]
